### PR TITLE
ua: deref call on reset_transp fail

### LINKED ISSUE #2531: closed call still exists
I have noticed that after baresip rejects incoming call like this:
```
04-15 11:44:29.013 15491 15541 D Baresip Lib: ua event (CALL_INCOMING) sip:test@test.tutpro.com
04-15 11:44:30.432 15491 15539 D Baresip Lib: reseting transports (1, 1)
04-15 11:44:30.435 15491 15542 D Baresip Lib: ua: ua_register sip:bar@test.tutpro.com
04-15 11:44:30.435 15491 15539 D Baresip Lib: ua event (REGISTERING) 
04-15 11:44:30.443 15491 15542 D Baresip Lib: call: rejecting incoming call from sip:test@test.tutpro.com (500 Transport of User Agent changed)
04-15 11:44:30.443 15491 15539 D Baresip Lib: ua event (CALL_CLOSED) Transport of User Agent changed
```
the call does not disappear, since when the next call comes in, my code
```
       case UA_EVENT_CALL_INCOMING:
            if (uag_call_count() > 1) {
                LOGD("auto-rejecting call from %s\n", prm);
                ua_hangup(ua, call, 486, "Busy Here");
```
auto-rejects the call.  

This only happens after `ua event (CALL_CLOSED) Transport of User Agent changed`.  Could it be that  in that case, the call is not properly closed, but remains in uag call list?

### PR DESCRIPTION
**Pull request by maximilianfridrich**
Fixes #2531.

### CODE REVIEW

File: src/uag.c [hunk 1]
```diff
@@ -757,6 +758,9 @@ int uag_reset_transp(bool reg, bool reinvite)
 					ua_event(ua, UA_EVENT_CALL_CLOSED,
 						 call, "Transport of "
 						 "User Agent changed");
+					lec_prev = lec->prev ? lec->prev : lec;
+					mem_deref(call);
+					lec = lec->next ? lec : lec_prev;
```
Comment (sreimers) [inline src/uag.c:763]: The common approach is to use a while loop, if destructor is used within `list` foreach itself:
```c
while (le) {
  call = lec->data;
  lec = lec->next;
  ...
}
```

**Comment by maximilianfridrich** [comment]
When I have one established call and one incoming call, and I call `uag_reset_transp(1, 1)` through a menu cmd, I get the following SEGV:

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007ff5bd68a452 in tcp_recv_handler (mb=0x55bff722b600, arg=0x55bff722e610)
    at /home/mfridrich/repos/bare/re/src/sip/transp.c:539
539			if (end <= conn->mb->end) {
[Current thread is 1 (Thread 0x7ff5bce17b80 (LWP 285011))]
(gdb) bt
#0  0x00007ff5bd68a452 in tcp_recv_handler (mb=0x55bff722b600, arg=0x55bff722e610)
    at /home/mfridrich/repos/bare/re/src/sip/transp.c:539
#1  0x00007ff5bd660167 in tcp_recv_handler (flags=1, arg=0x55bff756ee80) at /home/mfridrich/repos/bare/re/src/tcp/tcp.c:442
#2  0x00007ff5bd644901 in fd_poll (re=0x55bff6f61430) at /home/mfridrich/repos/bare/re/src/main/main.c:802
#3  0x00007ff5bd644d69 in re_main (signalh=0x55bff5aff444 <signal_handler>)
    at /home/mfridrich/repos/bare/re/src/main/main.c:976
#4  0x000055bff5affb67 in main (argc=3, argv=0x7fff12949ae8) at /home/mfridrich/repos/bare/baresip/src/main.c:323
```
Valgrind says:
```
==287555== Invalid read of size 8=396/1314 (bit/s) efps=31.8/19.1    
==287555==    at 0x495444B: tcp_recv_handler (transp.c:539)
==287555==    by 0x492A166: tcp_recv_handler (tcp.c:442)
==287555==    by 0x490E900: fd_poll (main.c:802)
==287555==    by 0x490ED68: re_main (main.c:976)
==287555==    by 0x10AB66: main (main.c:323)
==287555==  Address 0x1e3d2cd8 is 616 bytes inside a block of size 656 free'd
==287555==    at 0x484426F: free (vg_replace_malloc.c:884)
==287555==    by 0x491141E: mem_deref (mem.c:390)
==287555==    by 0x494D15A: destructor (msg.c:44)
==287555==    by 0x491137C: mem_deref (mem.c:376)
==287555==    by 0x4954446: tcp_recv_handler (transp.c:537)
==287555==    by 0x492A166: tcp_recv_handler (tcp.c:442)
==287555==    by 0x490E900: fd_poll (main.c:802)
==287555==    by 0x490ED68: re_main (main.c:976)
==287555==    by 0x10AB66: main (main.c:323)
==287555==  Block was alloc'd at
==287555==    at 0x4841888: malloc (vg_replace_malloc.c:393)
==287555==    by 0x4910D49: mem_alloc (mem.c:166)
==287555==    by 0x4910E9F: mem_zalloc (mem.c:198)
==287555==    by 0x4954932: tcp_connect_handler (transp.c:658)
==287555==    by 0x492A3BA: tcp_conn_handler (tcp.c:536)
==287555==    by 0x490E900: fd_poll (main.c:802)
==287555==    by 0x490ED68: re_main (main.c:976)
==287555==    by 0x10AB66: main (main.c:323)
==287555== 
==287555== Invalid read of size 8
==287555==    at 0x4954452: tcp_recv_handler (transp.c:539)
==287555==    by 0x492A166: tcp_recv_handler (tcp.c:442)
==287555==    by 0x490E900: fd_poll (main.c:802)
==287555==    by 0x490ED68: re_main (main.c:976)
==287555==    by 0x10AB66: main (main.c:323)
==287555==  Address 0xb5b5b5b5b5b5b5cd is not stack'd, malloc'd or (recently) free'd
==287555== 
==287555== 
==287555== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==287555==  General Protection Fault
==287555==    at 0x4954452: tcp_recv_handler (transp.c:539)
==287555==    by 0x492A166: tcp_recv_handler (tcp.c:442)
==287555==    by 0x490E900: fd_poll (main.c:802)
==287555==    by 0x490ED68: re_main (main.c:976)
==287555==    by 0x10AB66: main (main.c:323)
```
I have no idea yet why that happened. It looks like in `tcp_recv_handler` the `conn->mb` reference count (which is also referenced by `msg`) is zero in line 539, and the last dereferencing took place in line 537.

File: src/uag.c [hunk 2]
```diff
@@ -732,9 +732,10 @@ int uag_reset_transp(bool reg, bool reinvite)
 		if (!reinvite)
 			continue;
 
-
-		for (lec = ua_calls(ua)->head; lec; lec = lec->next) {
+		lec = ua_calls(ua)->head;
+		while (lec) {
 			struct call *call = lec->data;
+			lec = lec->next;
```
Comment (cspiel1) [inline src/uag.c:738]: Should we move this under the declaration block?

**Comment by sreimers** [comment]
I would assume that `mem_deref(msg);` should only deref its own reference and never destruct `conn`.

Untested:
```diff
diff --git a/src/sip/msg.c b/src/sip/msg.c
index 9abf1b3..babf708 100644
--- a/src/sip/msg.c
+++ b/src/sip/msg.c
@@ -41,7 +41,8 @@ static void destructor(void *arg)
        list_flush(&msg->hdrl);
        hash_flush(msg->hdrht);
        mem_deref(msg->hdrht);
-       mem_deref(msg->sock);
+       if (mem_nrefs(msg->sock) > 1)
+               mem_deref(msg->sock);
        mem_deref(msg->mb);
 }
```

But the other question is, why is `conn` already derefed?

**Comment by cspiel1** [comment]
The loop looks good now to me. But there is still a crash in `tcp_recv_handler`. So I set this to draft in order to avoid an early merge.

**Comment by juha-h** [comment]
I too tested diff:
```
diff --git a/src/uag.c b/src/uag.c
index a5bccf791..ee4f36405 100644
--- a/src/uag.c
+++ b/src/uag.c
@@ -732,9 +732,10 @@ int uag_reset_transp(bool reg, bool reinvite)
 		if (!reinvite)
 			continue;
 
-
-		for (lec = ua_calls(ua)->head; lec; lec = lec->next) {
+		lec = ua_calls(ua)->head;
+		while (lec) {
 			struct call *call = lec->data;
+			lec = lec->next;
 			struct stream *s;
 			const struct sa *raddr;
 			struct sa laddr;
@@ -757,6 +758,7 @@ int uag_reset_transp(bool reg, bool reinvite)
 					ua_event(ua, UA_EVENT_CALL_CLOSED,
 						 call, "Transport of "
 						 "User Agent changed");
+					mem_deref(call);
 					continue;
 				}
 				err = call_reset_transp(call, &laddr);
```
and now the reset transport crash is gone.  Don't know about `tcp_recv_handler`.

**Comment by sreimers** [comment]
Is there a way to reproduce the `tcp_recv_handler` crash with a test case? Looks like a regression that we should avoid when it is fixed.

**Comment by maximilianfridrich** [comment]
@sreimers I'll look into it. Currently, this fails for me every time, so merging this PR is not an option yet.

I'm still looking into the SEGV. I assume the tcp_recv_handler is called when the ACK message is received, and somehow the refcount of `conn->mb` is then zero after `mem_deref(msg)`.

**Comment by juha-h** [comment]
Maximilian Fridrich writes:

> @sreimers I'll look into it. Currently, this fails for me every time,
> so merging this PR is not an option yet. 

What kind of test case you have?

**Comment by maximilianfridrich** [comment]
@juha-h I am testing manually right now by making a single call from one baresip to another, and while the call is still in the INCOMING state, on the callee side I call `uag_reset_transp(1, 1)` through a menu cmd, as you suggested. I don't even need another ESTABLISHED call to get the SEGV.

**Comment by juha-h** [comment]
I made the same test.  `uag_reset_transp(1, 1)` didn't cause the crash, but next incoming call did it right away without any menu action:
```
baresip is ready.
test@test.tutpro.com: (prio 0) {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.7.0-0b01 (x86_64/linux))) [1 binding]
All 1 useragent registered successfully! (5 ms)
mwi: subscribing to messages for sip:test@test.tutpro.com
test@test.tutpro.com: selected for test-0x55a98f577010
ua: using connection-address 192.168.188.205 of SDP offer
sip:test@test.tutpro.com: Incoming call from: Bar sip:bar@test.tutpro.com - audio-video: sendrecv-inactive - (press 'a' to accept)
/entransp
call: rejecting incoming call from sip:bar@test.tutpro.com (500 Transport of User Agent changed)
test@test.tutpro.com: (prio 0) {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.7.0-0b01 (x86_64/linux))) [1 binding]
test@test.tutpro.com: selected for test-0x55a98f577010
ua: using connection-address 192.168.188.205 of SDP offer)
sip:test@test.tutpro.com: Incoming call from: Bar sip:bar@test.tutpro.com - audio-video: sendrecv-inactive - (press 'a' to accept)
Segmentation fault
```

**Comment by maximilianfridrich** [comment]
I'm getting closer. In `uag_reset_transp` at the very top it calls:
```
	/* Update SIP transports */
	sip_transp_flush(uag.sip);

	err = ua_transp_addall(net);
```
If I comment out those lines just for testing purposes, everything works fine. So it certainly has to do with us flushing these lists and hash tables in `transp.c:sip_transp_flush()`:
```
	hash_flush(sip->ht_conn);
	hash_flush(sip->ht_conncfg);
	list_flush(&sip->transpl);
```

**Comment by juha-h** [comment]
In my test, ACK does not cause the crash.  Crash happens when next income comes in:
```
10:40:23.601#
TCP 192.168.188.215:5060 -> 192.168.188.215:60412
INVITE sip:test-0x55a7358e8010@192.168.188.215:5020;transport=tcp SIP/2.0
Record-Route: <sip:192.168.188.215;transport=tcp;lr>
Via: SIP/2.0/TCP 192.168.188.215;branch=z9hG4bK0c64.e16bb4f8a989d4bd2539a177306fac4e.0;i=491
Via: SIP/2.0/TCP 192.168.188.205:34763;received=192.168.188.205;branch=z9hG4bKd41722f967033ada;rport=52350
Contact: <sip:bar@test.tutpro.com;gr=urn:uuid:6c56873f-06f3-761d-7e9d-56f5c2ddffd1>
Max-Forwards: 69
To: <sip:test@test.tutpro.com>
From: "Bar" <sip:bar@test.tutpro.com>;tag=aec8dd86a63621b2
Call-ID: fbeba7d59a8e3c7c
CSeq: 28775 INVITE
User-Agent: baresip v3.0.0 (aarch64/Android)
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,UPDATE,REFER
Supported: gruu,path,outbound,replaces,norefersub
Content-Type: application/sdp
Content-Length: 668
P-Asserted-Identity: <sip:+35892345673@test.tutpro.com>

v=0
o=- 300624349 1524260141 IN IP4 192.168.188.205
s=-
c=IN IP4 192.168.188.205
t=0 0
m=audio 2600 RTP/AVP 96 97 98 9 99 100 101 18 0 8 102
a=rtpmap:96 opus/48000/2
a=fmtp:96 stereo=0;sprop-stereo=0;maxaveragebitrate=28000;cbr=0;useinbandfec=1
a=rtpmap:97 AMR-WB/16000
a=rtpmap:98 AMR/8000
a=rtpmap:9 G722/8000
a=rtpmap:99 G7221/16000
a=fmtp:99 bitrate=32000
a=rtpmap:100 G726-24/8000
a=rtpmap:101 G726-16/8000
a=rtpmap:18 G729/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:102 telephone-event/8000
a=fmtp:102 0-15
a=sendrecv
a=rtcp-rsize
a=ssrc:1369005215 cname:sip:bar@test.tutpro.com
a=mid:0
a=minptime:20
a=ptime:20
a=label:1

test@test.tutpro.com: selected for test-0x55a7358e8010
ua: using connection-address 192.168.188.205 of SDP offer
stream: audio: starting mediaenc 'zrtp' (wait_secure=0)
stream: video: starting mediaenc 'zrtp' (wait_secure=0)
10:40:23.601#
TCP 192.168.188.215:60412 -> 192.168.188.215:5060
SIP/2.0 180 Ringing
Record-Route: <sip:192.168.188.215;transport=tcp;lr>
Via: SIP/2.0/TCP 192.168.188.215;branch=z9hG4bK0c64.e16bb4f8a989d4bd2539a177306fac4e.0;i=491
Via: SIP/2.0/TCP 192.168.188.205:34763;received=192.168.188.205;branch=z9hG4bKd41722f967033ada;rport=52350
To: <sip:test@test.tutpro.com>;tag=043d5d2e7c2b87f5
From: "Bar" <sip:bar@test.tutpro.com>;tag=aec8dd86a63621b2
Call-ID: fbeba7d59a8e3c7c
CSeq: 28775 INVITE
Server: baresip v3.0.0 (x86_64/Linux)
Contact: <sip:test@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce>
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,UPDATE,REFER
Content-Length: 0


sip:test@test.tutpro.com: Incoming call from: Bar sip:bar@test.tutpro.com - audio-video: sendrecv-inactive - (press 'a' to accept)
/entransp
call: rejecting incoming call from sip:bar@test.tutpro.com (500 Transport of User Agent changed)
10:40:29.581#
TCP 192.168.188.215:60412 -> 192.168.188.215:5060
SIP/2.0 500 Transport of User Agent changed
Via: SIP/2.0/TCP 192.168.188.215;branch=z9hG4bK0c64.e16bb4f8a989d4bd2539a177306fac4e.0;i=491
Via: SIP/2.0/TCP 192.168.188.205:34763;received=192.168.188.205;branch=z9hG4bKd41722f967033ada;rport=52350
To: <sip:test@test.tutpro.com>;tag=043d5d2e7c2b87f5
From: "Bar" <sip:bar@test.tutpro.com>;tag=aec8dd86a63621b2
Call-ID: fbeba7d59a8e3c7c
CSeq: 28775 INVITE
Server: baresip v3.0.0 (x86_64/Linux)
Content-Length: 0


10:40:29.588#
TCP 192.168.188.215:50178 -> 192.168.188.215:5060
REGISTER sip:test.tutpro.com SIP/2.0
Via: SIP/2.0/TCP 192.168.188.215:5020;branch=z9hG4bK55ebc7f732990efe;rport
Contact: <sip:test-0x55a7358e8010@192.168.188.215:5020;transport=tcp>;expires=0;+sip.instance="<urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce>";q=0.5;reg-id=1
Max-Forwards: 70
Route: <sip:192.168.188.215:5060;transport=tcp;lr>
To: <sip:test@test.tutpro.com>
From: "Test%20Account" <sip:test@test.tutpro.com>;tag=879a4319dd2c446d
Call-ID: fc03c4c73c09efdf
CSeq: 49778 REGISTER
User-Agent: baresip v3.0.0 (x86_64/Linux)
Supported: gruu, outbound, path
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,UPDATE,REFER
Content-Length: 0


10:40:29.589#
TCP 192.168.188.215:50178 -> 192.168.188.215:5060
REGISTER sip:test.tutpro.com SIP/2.0
Via: SIP/2.0/TCP 192.168.188.215:5020;branch=z9hG4bKee1c7574fc2fe719;rport
Contact: <sip:test-0x55a7358e8010@192.168.188.215:5020;transport=tcp>;expires=600;+sip.instance="<urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce>";q=0.5;reg-id=1
Max-Forwards: 70
Route: <sip:192.168.188.215:5060;transport=tcp;lr>
To: <sip:test@test.tutpro.com>
From: "Test%20Account" <sip:test@test.tutpro.com>;tag=82800403f0147c6f
Call-ID: cff7ff449b2da5c6
CSeq: 50556 REGISTER
User-Agent: baresip v3.0.0 (x86_64/Linux)
Supported: gruu, outbound, path
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,UPDATE,REFER
Content-Length: 0


10:40:29.589#
TCP 192.168.188.215:5060 -> 192.168.188.215:60412
ACK sip:test-0x55a7358e8010@192.168.188.215:5020;transport=tcp SIP/2.0
Via: SIP/2.0/TCP 192.168.188.215;branch=z9hG4bK0c64.e16bb4f8a989d4bd2539a177306fac4e.0;i=491
Max-Forwards: 69
To: <sip:test@test.tutpro.com>;tag=043d5d2e7c2b87f5
From: "Bar" <sip:bar@test.tutpro.com>;tag=aec8dd86a63621b2
Call-ID: fbeba7d59a8e3c7c
CSeq: 28775 ACK
Content-Length: 0


10:40:29.591#
TCP 192.168.188.215:5060 -> 192.168.188.215:50178
SIP/2.0 200 OK
Via: SIP/2.0/TCP 192.168.188.215:5020;branch=z9hG4bK55ebc7f732990efe;rport=50178;received=192.168.188.215
To: <sip:test@test.tutpro.com>;tag=4bde786d15ea2e19c68f4a0e9fc417d7.e9f88b49
From: "Test%20Account" <sip:test@test.tutpro.com>;tag=879a4319dd2c446d
Call-ID: fc03c4c73c09efdf
CSeq: 49778 REGISTER
Supported: outbound
Require: outbound
Server: OpenSIPg SIP Proxy (5.7.0-0b01 (x86_64/linux))
Content-Length: 0

test@test.tutpro.com: (prio 0) {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.7.0-0b01 (x86_64/linux))) [1 binding]
10:40:46.112#
TCP 192.168.188.215:5060 -> 192.168.188.215:50178
INVITE sip:test-0x55a7358e8010@192.168.188.215:5020;transport=tcp SIP/2.0
Record-Route: <sip:192.168.188.215;transport=tcp;lr>
Via: SIP/2.0/TCP 192.168.188.215;branch=z9hG4bK6df1.148ec23045bdf70699f215fb1b0d30d8.0;i=491
Via: SIP/2.0/TCP 192.168.188.205:34763;received=192.168.188.205;branch=z9hG4bKa6bbb2b36820c15f;rport=52350
Contact: <sip:bar@test.tutpro.com;gr=urn:uuid:6c56873f-06f3-761d-7e9d-56f5c2ddffd1>
Max-Forwards: 69
To: <sip:test@test.tutpro.com>
From: "Bar" <sip:bar@test.tutpro.com>;tag=c7d10b176c19f13f
Call-ID: 9e8ec456885c5f16
CSeq: 9723 INVITE
User-Agent: baresip v3.0.0 (aarch64/Android)
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,UPDATE,REFER
Supported: gruu,path,outbound,replaces,norefersub
Content-Type: application/sdp
Content-Length: 669
P-Asserted-Identity: <sip:+35892345673@test.tutpro.com>

v=0
o=- 2068774095 261271911 IN IP4 192.168.188.205
s=-
c=IN IP4 192.168.188.205
t=0 0
m=audio 24274 RTP/AVP 96 97 98 9 99 100 101 18 0 8 102
a=rtpmap:96 opus/48000/2
a=fmtp:96 stereo=0;sprop-stereo=0;maxaveragebitrate=28000;cbr=0;useinbandfec=1
a=rtpmap:97 AMR-WB/16000
a=rtpmap:98 AMR/8000
a=rtpmap:9 G722/8000
a=rtpmap:99 G7221/16000
a=fmtp:99 bitrate=32000
a=rtpmap:100 G726-24/8000
a=rtpmap:101 G726-16/8000
a=rtpmap:18 G729/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:102 telephone-event/8000
a=fmtp:102 0-15
a=sendrecv
a=rtcp-rsize
a=ssrc:1618728298 cname:sip:bar@test.tutpro.com
a=mid:0
a=minptime:20
a=ptime:20
a=label:1

test@test.tutpro.com: selected for test-0x55a7358e8010
ua: using connection-address 192.168.188.205 of SDP offer
stream: audio: starting mediaenc 'zrtp' (wait_secure=0)
stream: video: starting mediaenc 'zrtp' (wait_secure=0)
10:40:46.113#
TCP 192.168.188.215:50178 -> 192.168.188.215:5060
SIP/2.0 180 Ringing
Record-Route: <sip:192.168.188.215;transport=tcp;lr>
Via: SIP/2.0/TCP 192.168.188.215;branch=z9hG4bK6df1.148ec23045bdf70699f215fb1b0d30d8.0;i=491
Via: SIP/2.0/TCP 192.168.188.205:34763;received=192.168.188.205;branch=z9hG4bKa6bbb2b36820c15f;rport=52350
To: <sip:test@test.tutpro.com>;tag=4d999f7e9729dee1
From: "Bar" <sip:bar@test.tutpro.com>;tag=c7d10b176c19f13f
Call-ID: 9e8ec456885c5f16
CSeq: 9723 INVITE
Server: baresip v3.0.0 (x86_64/Linux)
Contact: <sip:test@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce>
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,UPDATE,REFER
Content-Length: 0

sip:test@test.tutpro.com: Incoming call from: Bar sip:bar@test.tutpro.com - audio-video: sendrecv-inactive - (press 'a' to accept)
Segmentation fault
```

**Comment by maximilianfridrich** [comment]
> I'm getting closer. In `uag_reset_transp` at the very top it calls:
> 
> ```
> 	/* Update SIP transports */
> 	sip_transp_flush(uag.sip);
> 	err = ua_transp_addall(net);
> ```
> 
> If I comment out those lines just for testing purposes, everything works fine. So it certainly has to do with us flushing these lists and hash tables in `transp.c:sip_transp_flush()`:
> 
> ```
> 	hash_flush(sip->ht_conn);
> 	hash_flush(sip->ht_conncfg);
> 	list_flush(&sip->transpl);
> ```
> 
> We're flushing these lists while we still have the incoming INVITE (server) transaction. After flushing, we send the 500, and for me it then crashes when we receive the ACK.

I think we have found the underlying issue. We have to think about how can handle existing transactions (that we abort) on transport change. I think if the transport had truly changed (unlike my fake test trigger), we would probably never receive the ACK. Nonetheless, this must be fixed.

**Comment by maximilianfridrich** [comment]
@juha-h Is there a way for you to test this without the manual cmd? I.e. an actual transport/IP change.

I suspect that my manual trigger is calling `uag_reset_transp` at a time when it should not be called. I think if it is called after an actual transport change, the old sockets will be gone anyways and we probably won't have this problem.

**Comment by juha-h** [comment]
> @juha-h Is there a way for you to test this without the manual cmd? I.e. an actual transport/IP change.
> 
> I suspect that my manual trigger is calling `uag_reset_transp` at a time when it should not be called. I think if it is called after an actual transport change, the old sockets will be gone anyways and we probably won't have this problem.

I can try with actual change using my Android app.  The app calls `uag_reset_transp` whenever there is an actual network change, e.g. when active network switches from WiFi to cellular.  But it is also called when there is change in inactive network, for example, when active network is cellular and WiFi is disconnected.

There should not be crash no matter when ``uag_reset_transp` is called.

**Comment by juha-h** [comment]
Call arrived over WiFi network to my baresip app and it started to ring.  Then I disconnected cellular network and my app called `uag_reset_transp`.  As result of that the ringing call was closed:
```
04-18 22:08:48.103  8022  8119 D Baresip Lib: ua event (CALL_CLOSED) Transport of User Agent changed
```
Then I made another call and baresip did not crash nor complain that there is already an active call.

What I'm wondering though is, why did `uag_reset_transp` cause closing of the ringing call, since that call didn't have anything to do with the disconnected cellular network.

When I did the same same test, but answered the ringing call and then disconnected the cellular network, the call stayed alive like it should.  So why didn't the ringing call stay alive?

**Comment by maximilianfridrich** [comment]
> What I'm wondering though is, why did `uag_reset_transp` cause closing of the ringing call, since that call didn't have anything to do with the disconnected cellular network.
> 
> When I did the same same test, but answered the ringing call and then disconnected the cellular network, the call stayed alive like it should. So why didn't the ringing call stay alive?

The reason is that if our transport/ip address changes, we have to tell the peer through a session update (re-INVITE or UPDATE), otherwise the peer will be sending packets to our old ip address and the communication is broken. If the call is in the incoming state, session UPDATEs are not allowed, **unless 100rel is used**. Note however that 100rel currently can still cause issues with forking INVITES, so be aware of that before enabling 100rel.

**Comment by juha-h** [comment]
I understand that, but in my test the transport used in the ringing call did not change.  It was the other (currently unused) transport that changed (was lost or became available).

Anyway, I'm OK with the PR as it now is.

**Comment by maximilianfridrich** [comment]
I guess this is due to how the netroam module and `uag_transp_reset` interact. If anything changes on the network interfaces, netroam calls `uag_transp_reset` which resets **all** transports. I have not changed this behavior.

**Comment by juha-h** [comment]
I'm not using netroam module in my Android app.  I don't remember the details anymore, but I think Android didn't provide what netroam needs.

There is this code in `uag_reset_transp`:
```
			s = audio_strm(call_audio(call));
			if (!s)
				s = video_strm(call_video(call));

			if (!s)
				continue;

			raddr = sdp_media_raddr(stream_sdpmedia(s));
			if (net_dst_source_addr_get(raddr, &laddr))
				continue;
```
If I read it correctly, it should skip the hangup since the source address is still found to the media address of the incoming INVITE, but for some reason the test does not match.

**Comment by maximilianfridrich** [comment]
Maybe we should open a separate issue for that, since the call deref seems to be fixed.

**Comment by juha-h** [comment]
Maximilian Fridrich writes:

> Maybe we should open a separate issue for that, since the call deref
> seems to be fixed.

Yes, the PR can be merged and I'll open a new issue for that.

**Pull request merged**